boring/src/aead.rs

```diff
@@ -0,0 +1,357 @@
+#[derive(Copy, Clone, PartialEq, Eq)]
+pub struct Aead(*const ffi::EVP_AEAD);
+
+impl Aead {
+    pub fn aes_128_gcm() -> Aead {
+        unsafe { Aead(ffi::EVP_aead_aes_128_gcm()) }
+    }
+
+    pub fn aes_192_gcm() -> Aead {
+        unsafe { Aead(ffi::EVP_aead_aes_192_gcm()) }
+    }
+
+    pub fn aes_256_gcm() -> Aead {
+        unsafe { Aead(ffi::EVP_aead_aes_256_gcm()) }
+    }
+
+    pub fn chacha20_poly1305() -> Aead {
+        unsafe { Aead(ffi::EVP_aead_chacha20_poly1305()) }
+    }
+
+    pub fn xchacha20_poly1305() -> Aead {
+        unsafe { Aead(ffi::EVP_aead_xchacha20_poly1305()) }
+    }
+
+    pub fn aes_128_ctr_hmac_sha256() -> Aead {
+        unsafe { Aead(ffi::EVP_aead_aes_128_ctr_hmac_sha256()) }
+    }
+
+    pub fn aes_256_ctr_hmac_sha256() -> Aead {
+        unsafe { Aead(ffi::EVP_aead_aes_256_ctr_hmac_sha256()) }
+    }
+
+    pub fn aes_128_gcm_siv() -> Aead {
+        unsafe { Aead(ffi::EVP_aead_aes_128_gcm_siv()) }
+    }
+
+    pub fn aes_256_gcm_siv() -> Aead {
+        unsafe { Aead(ffi::EVP_aead_aes_256_gcm_siv()) }
+    }
+
+    /// Returns the length of keys used with this AEAD.
+    pub fn key_len(&self) -> usize {
+        unsafe { EVP_AEAD_key_length(self.0) as usize }
+    }
+
+    /// Returns the length of nonces used with this AEAD.
+    pub fn nonce_len(&self) -> usize {
+        unsafe { EVP_AEAD_nonce_length(self.0) as usize }
+    }
+
+    /// Returns the maximum ciphertext overhead with this AEAD.
+    pub fn max_ciphertext_overhead(&self) -> usize {
+        unsafe { EVP_AEAD_max_overhead(self.0) as usize }
+    }
+
+    /// Returns the maximum tag length with this AEAD.
+    pub fn max_tag_length(&self) -> usize {
+        unsafe { EVP_AEAD_max_tag_len(self.0) as usize }
+    }
+
+    pub fn as_ptr(&self) -> *const ffi::EVP_AEAD {
+        self.0
+    }
+}
+
+unsafe impl Sync for Aead {}
+unsafe impl Send for Aead {}
+
+/// Represents an AEAD context.
+pub struct AeadCrypter {
+    ctx: *mut ffi::EVP_AEAD_CTX,
+    max_overhead: usize,
+}
+
+impl AeadCrypter {
+    /// Creates a new `AeadCrypter`.
+    pub fn new(aead: Aead, key: &[u8]) -> Result<AeadCrypter, ErrorStack> {
+        ffi::init();
+
+        unsafe {
+            let ctx = cvt_p(EVP_AEAD_CTX_new(
+                aead.as_ptr(),
+                key.as_ptr(),
+                key.len(),
+                0, // supply tag_len = 0 so it picks the default one.
+            ))?;
+            let aeadcrypter = AeadCrypter {
+                ctx,
+                max_overhead: aead.max_ciphertext_overhead(),
+            };
+            Ok(aeadcrypter)
+        }
+    }
+
+    /// Seals data by encrypting and authenticating the provided data and authenticating the provided aad.
+    ///
+    /// Returns the number of bytes written to output.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `output.len() > c_int::max_value()`.
+    ///
+    /// Panics if `output.len() < data.len() + aead.max_overhead()`.
```

Comment:
I know this mimics the interface of the underlying library, but can we provide a better one here? I think it's fine to allocate a new buffer instead of taking an `output` argument - that matches the behavior of [`DH::params_to_pem`](https://docs.rs/boring/latest/boring/dh/struct.Dh.html#method.params_to_pem).

Reply:
I agree, I was trying to match `symm.rs`.

https://github.com/cloudflare/boring/blob/e6ddc403337de031e446607cd161bcc41f2610db/boring/src/symm.rs#L446-L482